Add community posts from discussions (#4058)
woodpecker-ci/woodpecker: docs/blog/2023-12-15-podman-sigstore/index.md

---
title: '[Community] Podman image build with sigstore'
description: Build images in Podman with sigstore signature checking and signing
slug: podman-image-build-sigstore
authors:
  - name: handlebargh
    url: https://github.com/handlebargh
    image_url: https://github.com/handlebargh.png
hide_table_of_contents: false
tags: [community, image, podman, sigstore, signature]
---

<!-- cspell:ignore BQVUJ Containerfile cosing distroless fulcio keypair nonroot QVRFLS rekor skopeo -->

This example shows how to build a container image with podman while verifying the base image and signing the resulting image.

> The image being pulled uses a keyless signature while the image being built will be signed by a pre-generated private key.

Reviewer: ```suggestion
The image being pulled uses a keyless signature, while the image being built will be signed by a pre-generated private key.
```

## Prerequisites

### Generate signing keypair

You can use cosing or skopeo to generate the keypair.

Using skopeo:

```bash
skopeo generate-sigstore-key --output-prefix myKey
```

This command will generate a `myKey.private` and a `myKey.pub` keyfile.

> Store the `myKey.private` as secret in Woodpecker. In the example below the secret is called `sigstore_private_key`

Reviewer: ```suggestion
Store the `myKey.private` as secret in Woodpecker. In the example below, the secret is called `sigstore_private_key`
```

```nohighlight
.
├── Containerfile
├── main.go
├── go.mod
├── go.sum
├── .woodpecker.yml
└── Makefile
```

### Containerfile

The Containerfile refers to the base image that will be verified when pulled.

```dockerfile
FROM gcr.io/distroless/static-debian12:nonroot
COPY app /app
CMD ["/app"]
```

### Woodpecker workflow

```yaml
steps:
  build:
    image: docker.io/library/golang:1.21
    pull: true
    commands:
      - make build

  publish:
    image: quay.io/podman/stable:latest
    # Caution: This image is built daily. It might fill up your image store quickly.
    pull: true
    # Fill in the trusted checkbox in Woodpecker's settings as well
    privileged: true
    commands:
      # Configure podman to use sigstore attachments for both, the registry you pull from and the registry you push to.
      - |
        printf "docker:
          registry.gitlab.com:
            use-sigstore-attachments: true
          gcr.io:
            use-sigstore-attachments: true" >> /etc/containers/registries.d/default.yaml

      # At pull, check the keyless sigstore signature of the distroless image.

>       # This is a very strict container policy. It allows to pull from gcr.io/distroless only. Every other registry will be rejected.

Reviewer: ```suggestion
      # This is a very strict container policy. It allows pulling from gcr.io/distroless only. Every other registry will be rejected.
```
